src/protobuf.py

```diff
@@ -49,15 +51,45 @@
         n = shifted
 
 
-class UVarintType:
-    WIRE_TYPE = 0
+# protobuf interleaved signed encoding:
+# https://developers.google.com/protocol-buffers/docs/encoding#structure
+# the idea is to save the sign in LSbit instead of twos-complement.
+# so counting up, you go: 0, -1, 1, -2, 2, ... (as the first bit changes, sign flips)
+#
+# To achieve this with a twos-complement number:
+# 1. shift left by 1, leaving LSbit free
+# 2. if the number is negative, do bitwise negation.
+#    This keeps positive number the same, and converts negative from twos-complement
+#    to the appropriate value, while setting the sign bit.
+#
+# The original algorithm makes use of the fact that arithmetic (signed) shift
+# keeps the sign bits, so for a n-bit number, (x >> n) gets us "all sign bits".
+# Then you can take "number XOR all-sign-bits", which is XOR 0 (identity) for positive
+# and XOR 1 (bitwise negation) for negative. Cute and efficient.
+#
+# But this is harder in Python because we don't natively know the bit size of the number.
+# So we have to branch on whether the number is negative.
+
+def sint_to_uint(sint):
+    res = sint << 1
+    if sint < 0:
+        res = ~res
```

Comment:
Will this work correctly for both sizes (32 and 64 bit?).

I would love to see some unit tests for these.

Reply:
If there are tests in python-trezor, then that's enough. Thanks!

Reply:
here are the tests, for the record
https://github.com/trezor/python-trezor/blob/master/trezorlib/tests/unit_tests/test_protobuf.py